pkg/agent/multicast/mcast_route.go

```diff
@@ -0,0 +1,455 @@
+	mc.Origin = [4]byte{origin[0], origin[1], origin[2], origin[3]}
+	g := group.To4()
+	mc.Mcastgrp = [4]byte{g[0], g[1], g[2], g[3]}
+	ttls := [32]byte{}
+	for _, v := range oifVIFs {
+		ttls[v] = 1
+	}
+	mc.Ttls = ttls
+	mc.Parent = iif
+	return multicastsyscall.SetsockoptViMfcctl(c.sockFD, syscall.IPPROTO_IP, multicastsyscall.MRT_ADD_MFC, mc)
+}
+
+func (c *MRouteClient) delMrouteEntry(src net.IP, group net.IP, iif uint16) (err error) {
+	mc := &multicastsyscall.Mfcctl{}
+	origin := src.To4()
+	mc.Origin = [4]byte{origin[0], origin[1], origin[2], origin[3]}
+	g := group.To4()
+	mc.Mcastgrp = [4]byte{g[0], g[1], g[2], g[3]}
+	mc.Parent = iif
+	return multicastsyscall.SetsockoptViMfcctl(c.sockFD, syscall.IPPROTO_IP, multicastsyscall.MRT_DEL_MFC, mc)
+}
+
+func (c *MRouteClient) flushMRoute() {
+	klog.InfoS("Clearing multicast routing table entries")
+	err := multicastsyscall.SetsockoptVifctl(c.sockFD, syscall.IPPROTO_IP, multicastsyscall.MRT_FLUSH, &multicastsyscall.Vifctl{})
+	if err != nil {
+		klog.InfoS("Failed to clear multicast routes")
+	}
+}
+
+func (c *MRouteClient) deleteInboundMrouteEntryByGroup(group net.IP) (err error) {
+	klog.V(2).InfoS("Deleting multicast group", "group", group)
+	mEntries, _ := c.inboundRouteCache.ByIndex(GroupNameIndexName, group.String())
+	for _, route := range mEntries {
+		entry := route.(*inboundMulticastRouteEntry)
+		err := c.delMrouteEntry(net.ParseIP(entry.src), net.ParseIP(entry.group), entry.vif)
+		if err != nil {
+			return err
+		}
+		c.inboundRouteCache.Delete(route)
+	}
+	return nil
+}
+
+// addOutboundMrouteEntry configures multicast route from Antrea gateway to oifs(outbound multicast interfaces)
+// to allow multicast sender Pods send multicast traffic to external.
+func (c *MRouteClient) addOutboundMrouteEntry(src net.IP, group net.IP, oifs []string) (err error) {
+	klog.InfoS("Adding outbound multicast route", "src", src, "group", group, "oifs", oifs)
+	vifs := make([]uint16, 0, len(oifs))
+	for _, oif := range oifs {
+		vif, err := c.vifAllocator.getVIF(oif)
+		if err != nil {
+			return err
+		}
+		vifs = append(vifs, vif)
+	}
+
+	inboundVIF, _ := c.vifAllocator.getVIF(c.gatewayInterface)
+	err = c.addMrouteEntry(src, group, uint16(inboundVIF), vifs)
+	if err != nil {
+		return err
+	}
+	routeEntry := &outboundMulticastRouteEntry{
+		group: group.String(),
+		src:   src.String(),
+		vifs:  vifs,
+	}
+	_, exist, _ := c.outboundRouteCache.GetByKey(routeEntry.group + "/" + routeEntry.src)
+	if !exist {
+		c.outboundRouteCache.Add(routeEntry)
+	} else {
+		c.outboundRouteCache.Update(routeEntry)
+	}
+	return nil
+}
+
+// addInboundMrouteEntry configures multicast route from external interface to Antrea gateway
+// to allow multicast receiver Pods receive multicast traffic from external.
+func (c *MRouteClient) addInboundMrouteEntry(src net.IP, group net.IP, inboundVIF uint16) (err error) {
+	klog.InfoS("Adding inbound multicast route", "src", src, "group", group, "inboundVIF", inboundVIF)
```

Comment:
Do you think we should check if we have a multicast receiver in the local Pods first, and only forward those traffic targeting to Pod receivers to antrea-gw0?

Reply:
We need to add this check. Implement this check with `groupCache`

Reply:
updated